### ISSUE TITLE
Do not remap TAB in cmd line if cmdline is not used

### ISSUE THREAD

**Issue by kristijanhusak** [hrsh7th/nvim-cmp#436]
**Describe the bug**
Hi,

recently you added some support for cmdline autocompletion. I gave it a try, but It's just not working for me, so I removed it.
I'm having a small search helper that allows me to search things strictly or by regex, and I toggle that search more with `<tab>` in cmd line. You can see it [here](https://github.com/kristijanhusak/neovim-config/blob/master/nvim/lua/partials/search.lua#L59).
This stopped working for me because cmp maps `<tab>` in cmdline, even if I don't use any of the cmdline features of cmp.

Could this mapping somehow not be set if cmdline feature is not used?

**Minimal config based on [this](https://github.com/hrsh7th/nvim-cmp/blob/main/utils/vimrc.vim)**

```lua
local cmp = require'cmp'
cmp.setup({
  sources = {
    { name = 'nvim_lsp' },
  },
})

```

**To Reproduce**

1. Open Neovim
2. do `:cmap <TAB>`
3. See that it's mapped to something like this:
   `c  <Tab>       *@<Cmd>call v:lua.cmp.utils.keymap.listen.run(14916799)<CR>  `

**Expected behavior**
I expect not to have `tab` mapped to anything if i'm not using cmdline autocompletion.

**Additional context**

**Comment by hrsh7th** [comment]
It's difficult due to nvim-cmp's design...

I hope my keymap fallbacking mechanism supports your `<tab>` mapping in cmdline mode.

Could you create minimal broken mapping?

I just tried to check the following mapping but it's not working even if disabled nvim-cmp. (I removed nvim-cmp from runtimepath for checking).

```
if has('vim_starting')
  set encoding=utf-8
endif
scriptencoding utf-8

if &compatible
  set nocompatible
endif

let s:plug_dir = expand('/tmp/plugged/vim-plug')
if !filereadable(s:plug_dir .. '/plug.vim')
  execute printf('!curl -fLo %s/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim', s:plug_dir)
end

execute 'set runtimepath+=' . s:plug_dir
call plug#begin(s:plug_dir)
" Plug 'hrsh7th/nvim-cmp'
" Plug 'hrsh7th/cmp-buffer'
call plug#end()

set indentkeys+==end
set cindent

lua <<EOF
local is_toggle = true
_G.search = {}

function search.toggle_search_mode()
  is_toggle = true
  mode = mode == 'regex' and 'term' or 'regex'

  return vim.fn.getcmdline()
end

vim.api.nvim_set_keymap('c', '<tab>', '<C-\\>ev:lua.search.toggle_search_mode()<CR><CR>', { noremap = false, nowait = true, silent = true })


-- local cmp = require "cmp"
-- cmp.setup {
--   mapping = {
--     ['<CR>'] = cmp.mapping.confirm({ select = true }),
--     ['<tab>'] = cmp.mapping(function(fallback)
--       if cmp.visible() then
--         cmp.select_next_item()
--       else
--         fallback()
--       end
--     end, { 'i',  's' }),
--   },
--   sources = {
--     { name = "buffer" },
--   },
-- }
EOF

```

**Comment by hrsh7th** [comment]
@kristijanhusak I hope to fix this fallback bug but you can use the following workaround.

```lua
cmp.setup {
  mapping = {
    ['<Tab>'] = cmp.config.disable,
    ['<S-Tab>'] = cmp.config.disable,
  }
}
```

**Comment by kristijanhusak** [comment]
@hrsh7th does this disable `<TAB>` completely? I'm using tab for regular autocompletion.
Here is minvimrc that reproduces it:
```vim
if has('vim_starting')
  set encoding=utf-8
endif
scriptencoding utf-8

if &compatible
  set nocompatible
endif

let mapleader = ','
let s:plug_dir = expand('/tmp/plugged/vim-plug')
if !filereadable(s:plug_dir .. '/plug.vim')
  execute printf('!curl -fLo %s/autoload/plug.vim --create-dirs https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim', s:plug_dir)
end

set packpath=/tmp/plugged/pack
execute 'set runtimepath+=' . s:plug_dir
call plug#begin(s:plug_dir)
" Plug 'hrsh7th/nvim-cmp'
" Plug 'hrsh7th/cmp-buffer'
call plug#end()

lua <<EOF
local is_toggle = true
local mode = 'term'
_G.search = {}

function search.toggle_search_mode()
  is_toggle = true
  mode = mode == 'regex' and 'term' or 'regex'

  return vim.fn.getcmdline()
end

function search.run()
  local term = ''
  vim.api.nvim_set_keymap('c', '<tab>', '<C-\\>ev:lua.search.toggle_search_mode()<CR><CR>', { noremap = false, nowait = true, silent = true })
  local status, t = pcall(vim.fn.input, 'Enter '..mode..': ', term)

  term = t

  if is_toggle then
    is_toggle = false
    return search.run(term)
  end
end

vim.api.nvim_set_keymap('n', '<leader>f', ':call v:lua.search.run()<CR>', { noremap = false, nowait = true, silent = true })


--local cmp = require "cmp"
-- cmp.setup {
--   sources = {
--     { name = "buffer" },
--   },
-- }
EOF
```
Steps
* Run this with `nvim -u mininit.vim`.
* Once started, make sure `:cmap <tab>` returns "no mapping found".
* execute `,f` to start search
* press tab, it changes from `Enter term` to `Enter regex` and vice versa

To reproduce bug:
* Uncomment `Plug` part, no need for cmp.setup
* Try running Steps again, see that Tab is not working.

**Comment by hrsh7th** [comment]
OK. I found the cause.
Your mapping uses `global mapping` and it initialize lazily. So my keymap utility seems always find my own utilities mapping...

I will consider how to fix it...

**Comment by wookayin** [comment]
Thanks for fixing this issue. I found that this patch (e8cb695) somehow fixes a mysterious bug where `<TAB>` in the command line mode suddenly stopped working for simple completions starting with some special characters, such as (`:e ~/.vim/...`). Great catch and fix!

**Comment by hrsh7th** [comment]
Sorry. I think it isn't fixed yet...

**Comment by kristijanhusak** [comment]
@hrsh7th first issue is fixed (changing between term and regex), but then path completion fails on regular `input('name:', '', 'file')`. As a workaround, I added `vim.cmd[[cunmap <buffer> <TAB>]]` that removes your mapping before the input appears. That solves the issue for this one case, but I think that every input prompt autocompletion is now broken.

**Comment by hrsh7th** [comment]
@kristijanhusak Could you test with https://github.com/hrsh7th/nvim-cmp/pull/452 if you have time?

I mostly rewrited the keymap.lua. It now handles the global/buffer mapping correctly. (Maybe).

**Comment by kristijanhusak** [comment]
@hrsh7th I tested it, it works perfectly!

**Comment by hrsh7th** [comment]
First of all, I'm sorry to take your time due to a issue.
This report inspired me to rethink keymap.lua. Thank you very much.

**Comment by kristijanhusak** [comment]
No worries, thank you for fixing it.

**Comment by wookayin** [comment]
Memo: the fix made in #452 still "maps" Tab, Shift-tab, etc. but simply bypasses if nvim-cmp didn't configure cmdline. (Different from what was asked in this issue as the title)
